src/source/Ice/NatBehaviorDiscovery.c

```diff
@@ -0,0 +1,404 @@
+/**
+ * Kinesis Video Producer Ice Utils
+ */
+#define LOG_CLASS "NatTypeDiscovery"
+#include "../Include_i.h"
+
+#define DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT                  5
+#define DEFAULT_TEST_NAT_TEST_RESPONSE_WAIT_TIME                    500 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND
+
+#define NAT_BEHAVIOR_NONE_STR                                       (PCHAR) "NONE"
+#define NAT_BEHAVIOR_NOT_BEHIND_ANY_NAT_STR                         (PCHAR) "NOT_BEHIND_ANY_NAT"
+#define NAT_BEHAVIOR_NO_UDP_CONNECTIVITY_STR                        (PCHAR) "NO_UDP_CONNECTIVITY"
+#define NAT_BEHAVIOR_ENDPOINT_INDEPENDENT_STR                       (PCHAR) "ENDPOINT_INDEPENDENT"
+#define NAT_BEHAVIOR_ADDRESS_DEPENDENT_STR                          (PCHAR) "ADDRESS_DEPENDENT"
+#define NAT_BEHAVIOR_PORT_DEPENDENT_STR                             (PCHAR) "PORT_DEPENDENT"
+
+typedef struct {
+    /* Should be able to contain max number of binding response we can get */
+    PStunPacket bindingResponseList[DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT * 6];
+    UINT32 bindingResponseCount;
+    CVAR cvar;
+    MUTEX lock;
+} NatTestData, *PNatTestData;
+
+STATUS natTestIncomingDataHandler(
+        UINT64 customData, PSocketConnection pSocketConnection, PBYTE pBuffer, UINT32 bufferLen,
+        PKvsIpAddress pSrc, PKvsIpAddress pDest)
+{
+    UNUSED_PARAM(pSocketConnection);
+    UNUSED_PARAM(pSrc);
+    UNUSED_PARAM(pDest);
+
+    STATUS retStatus = STATUS_SUCCESS;
+    PNatTestData pNatTestData = (PNatTestData) customData;
+    PStunPacket pStunPacket = NULL;
+
+    MUTEX_LOCK(pNatTestData->lock);
+    if (pNatTestData->bindingResponseCount < DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT * 3) {
+        CHK_STATUS(deserializeStunPacket(pBuffer, bufferLen, NULL, 0, &pStunPacket));
+        pNatTestData->bindingResponseList[pNatTestData->bindingResponseCount++] = pStunPacket;
+    }
+    MUTEX_UNLOCK(pNatTestData->lock);
+    CVAR_SIGNAL(pNatTestData->cvar);
+
+CleanUp:
+
+    return STATUS_SUCCESS;
+}
+
+STATUS executeNatTest(PStunPacket bindingRequest, PKvsIpAddress pDestAddr, PSocketConnection pSocketConnection,
```

Comment:
Would prefer to remove the word "test" - something simple like discoverNatBehavior or something like that

Reply:
this mirrors the "test" described in the rfc: https://tools.ietf.org/html/rfc5780#section-4.3

Reply:
Oh ok